`storage`: set cached disk usage after compaction rounds

WillemKauf opened the pull request:
# Prelude
[Check out this cool flamegraph!](https://pprof.me/94592b10a8121eb1844e4e99e92b2dc2/?cur_path=%25255B%25255D)

What's not so cool is the nearly 15% of cumulative CPU time spent within `segment::persistent_size()`:

![image](https://github.com/user-attachments/assets/fdec4082-a7b3-4959-99ba-a2fb0bf2da0b)

# The Cover Letter

The major motivation for these changes is the number of calls being made to `fstat()` for each `segment` during a housekeeping round. These syscalls must go through the seastar `syscall_work_queue`, and take up much more CPU utilization than they should when studying a flamegraph.

The common thread in all of the flamegraph stack traces is `segment::persistent_size()`, which issues `fstat()` commands to acquire the size of the `segment`, `segment_index` and `compacted_index` if they are not currently cached. Compaction on a segment unsets all of these via `segment::clear_cached_disk_usage()`:

https://github.com/redpanda-data/redpanda/blob/00c76c3aa6a9815a84e14c2f9ad1ba6d7bde02b3/src/v/storage/segment_utils.cc#L520 

Within `log_manager::gc_loop()`, we query a `log` for its `disk_usage()`, which must make calls to EACH `segment`'s `persistent_size()`. What is likely even more responsible for the calls to `disk_usage()` is from the `disk_space_manager::run_loop()`, which according to the default `retention_local_trim_interval`, occurs every `30s`.

https://github.com/redpanda-data/redpanda/blob/00c76c3aa6a9815a84e14c2f9ad1ba6d7bde02b3/src/v/resource_mgmt/storage.cc#L515

It is clear to see that compaction constantly invalidating these values and then not re-assigning them could result in a lot of time spent in syscalls to disk every time urgent gc is issued/every space management loop. Thankfully, we have enough state in memory to reassign (at the very least, a rough estimation) the size instead of just clearing the cache.

This should all have the intended effect of cutting down on system calls in the housekeeping layer.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Improvements

* Cut down the amount of time spent in `fstat()` syscalls during `storage` layer housekeeping.

WillemKauf committed:
`storage`: add `compacted_index_writer::size_bytes()`

A utility function that calls into the underlying
`segment_appender->size_bytes()`.

This will be used in future commits as a way to get the size of a
`.compaction_index` file on disk without extra calls to `stat()`.

WillemKauf committed:
`storage`: add `segment::set_cached_disk_usage()`

Sets the cached disk usage for the `segment` and `compacted_index`. These
are usually the values from the underlying `segment_{appender/reader}`
for either of these objects. The base `segment_index` sets its disk usage
in `flush_to_file()`.

This function is used within the compaction subsystem to cut down on
future calls to `stat()`, since the cached disk usage is cleared during
every round of compaction, and we know from in-memory state how large
these objects are going to be on disk.

WillemKauf committed:
`storage`: set `_disk_usage_size` in `segment_index` at new sites

After this change, the `_disk_usage_size` will be set at two new locations:
1. `materialize_index_from_file()`, where we get the file size before reading
   the `.base_index` file.
2. `flush_to_file()`, where we iterate over the fragments of the `serde`'d
   `index_state` and write them to an output buffer.

These new assignments serve to cut down on the number of times we must
perform a call to `ss::file_size()` within `segment_index:disk_usage()`.

WillemKauf committed:
`storage`: add `segment_reader::fsize()`

Currently, `segment_reader::stat()` is only called in places where
we require the file size of the underlying `segment` file.

To offer a way of getting the size of a file on disk without needing to
issue a syscall which returns all of the `stat(2)` data, add `fsize()`.

WillemKauf committed:
`storage`: use `segment_reader::fsize()` instead of `stat()`

These locations only called `stat()` to obtain `st_size`. Use `fsize()`
instead.

WillemKauf committed:
`storage`: co-routinize `write_clean_compacted_index()`

No functional changes. Exception handling should remain the same here
in co-routinized version due to `try/catch` blocks and capturing of
exception with `eptr`.

WillemKauf committed:
`storage`: add `cmp_idx_size_after` to `compaction_result`

As a way to pass the new size of the rewritten `.compaction_index` file
between functions in the compaction util call stack.

This will be used in a future commit to manually set the cached size
of the `.compaction_index` within a `segment` in memory.

WillemKauf committed:
`storage`: change return type of `do_self_compact_segment()` to `compaction_result`

No functional changes here, `self_compact_segment()` returns a `compaction_result`
using the same data that could be used to build the object at a layer down
in the call stack.

This will also have the added benefit of allowing the `cmp_idx_size_after`
to be passed back to callers of `self_compact_segment()`.

WillemKauf committed:
`storage`: return `cmp_idx_size` from various compaction functions

Bubble up the size of the new `.compaction_index` from within the call
stack of:

```
* do_self_compact_segment()
  * do_compact_segment_index()
    * write_clean_compacted_index()
      * do_write_clean_compacted_index()
        * copy_filtered_entries()
```

for use in `do_self_compact_segment()`.

WillemKauf committed:
`storage`: set cached disk usage in `do_swap_data_file_handles()`

The major motivation for these changes is the number of calls being made
to `fstat()` for each `segment` during a housekeeping round. These syscalls
must go through the seastar `syscall_work_queue`, and take up much more
CPU utilization than they should when studying a flamegraph.

The common thread in all of the flamegraph stack traces is
`segment::persistent_size()`, which issues `fstat()` commands to acquire
the size of the `segment`, `segment_index` and `compacted_index` if they
are not currently cached.

We issue calls to this function from within `log_manager::gc_loop()`, which
queries a `log` for its `disk_usage()`, which must make calls to EACH
`segment`'s `persistent_size()`.

It is clear to see that compaction constantly invalidating these values
and then not re-assigning them could result in a lot of time spent in syscalls
to disk. Thankfully, we have enough state in memory to reassign (at the very least,
a rough estimation) the size instead of just clearing the cache.

After plumbing the returned value of `new_cmp_idx_size` throughout
the compaction utils layer, we can now call `s->set_cached_disk_usage()`
after clearing it within `do_swap_data_file_handles()`. The size of the `segment`
is set according to the `segment_reader::file_size()`, which is guaranteed to
be correct, as a call to `segment_reader::load_size()` is issued right above it
in the function. The `segment_index`'s size will be assigned instead when a call
to `segment_index::flush()` is made.

This should all have the intended effect of cutting down on system calls in the
housekeeping layer.

WillemKauf committed:
`storage`: remove call to `clear_cached_disk_usage()`

This call is unnecessary. If the `segment` was indeed compacted,
its cached disk usage would have been already cleared within
`do_swap_data_file_handles()`.

Keeping this call would now erroneously remove the newly cached
values we just set.

WillemKauf committed:
`storage`: add `segment_accessor`

This will be used in a future test which checks the values of
the cached disk usages.

WillemKauf committed:
`storage`: add `segment_cached_disk_usage_set_after_compaction`

To ensure that these values are indeed set, and not cleared,
after various forms of compaction, thereby saving us on calls to
`file_stat()` during future housekeeping rounds.

dotnwat commented:
will we be able to get a flamegraph before/after this PR?

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/67586-->
<!--85ddf9bf-c52f-4c81-ac9f-e9592f540f10-->
<!--title-->
## Retry command for [Build#67586](https://buildkite.com/redpanda/redpanda/builds/67586)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/topic_delete_test.py::TopicDeleteStressTest.stress_test
```
<!--/body-->

WillemKauf commented:
>will we be able to get a flamegraph before/after this PR?

That would be cool, the flamegraph in the OP was from a dev cluster in the wild, but I could do some local testing to try to get a flame graph targeting compaction with these changes to gauge the benefit

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--a87e7b11-2424-455d-ac82-533a5cc248ef-->
<!--TEST RESULTS-->
<!--535c2dde-4fce-45a9-8040-ff4d5df38d25-->
<!--TEST RESULTS-->
<!--bd3d46fa-bca4-4da4-a393-7a2fb1981ba3-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67598">build#67598</a></summary>


| test_class          | test_method      | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:--------------------|:-----------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeMetricsTest | test_lag_metrics | {"cloud_storage_type": 1} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67598#01978631-ef5d-4a28-a644-9c769b1c6ccc | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67620">build#67620</a></summary>


| test_class                   | test_method                             | test_arguments                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-----------------------------|:----------------------------------------|:-------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest    | test_read_chunks                        |                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67620#019788b8-52db-4d0c-bbd6-494d87e8197d | FLAKY         | 20/21    | upstream reliability is '98.28080229226362'. current run reliability is '95.23809523809523'. drift is 3.04271 and the allowed drift is set to 50. The test should PASS  |
| CloudStorageTimingStressTest | test_cloud_storage_with_partition_moves | {"cleanup_policy": "compact,delete"}                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67620#019788d5-1152-41ba-b521-462060513ab0 | FLAKY         | 18/21    | upstream reliability is '98.00664451827242'. current run reliability is '85.71428571428571'. drift is 12.29236 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest        | test_migrated_topic_data_integrity      | {"params": {"cancellation": {"dir": "in", "stage": "executing"}, "use_alias": false}, "transfer_leadership": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67620#019788b8-52db-4d0c-bbd6-494d87e8197d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| IcebergUsageTest             | test_iceberg_usage                      | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"}                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67620#019788d5-114f-43f5-b9ef-a293f4d51cca | FLAKY         | 20/21    | upstream reliability is '85.7707509881423'. current run reliability is '95.23809523809523'. drift is -9.46734 and the allowed drift is set to 50. The test should PASS  |
| DisablingPartitionsTest      | test_disable                            |                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67620#019788b8-52d9-4898-9be0-a48edebbea31 | FLAKY         | 16/21    | upstream reliability is '93.32161687170475'. current run reliability is '76.19047619047619'. drift is 17.13114 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67836">build#67836</a></summary>


| test_class                         | test_method                   | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------------------|:------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67836#0197a54a-fc52-4c2d-8a75-096d218d5484 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| IcebergUsageTest                   | test_iceberg_usage            | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "spark"}                                                       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67836#0197a54e-9c2e-415a-b002-d106cddcbd37 | FLAKY         | 18/21    | upstream reliability is '86.99633699633699'. current run reliability is '85.71428571428571'. drift is 1.28205 and the allowed drift is set to 50. The test should PASS |
| PartitionForceReconfigurationTest  | test_basic_reconfiguration    | {"acks": 1, "controller_snapshots": false, "restart": true}                                                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67836#0197a54e-9c32-45f1-a07b-7e3d9e2f7ef6 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest           | test_node_operations          | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67836#0197a54a-fc4f-4e50-8ad7-1e2506b975b3 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| DisablingPartitionsTest            | test_disable                  |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67836#0197a54e-9c31-4e78-ace7-e87ea3e42253 | FLAKY         | 18/21    | upstream reliability is '93.11594202898551'. current run reliability is '85.71428571428571'. drift is 7.40166 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

WillemKauf commented on src/v/storage/segment_utils.cc:
```diff
@@ -238,16 +238,20 @@ ss::future<> copy_filtered_entries(
               model::no_timeout)
             // must be last
-            .finally([&writer] {
+            .then_wrapped([&writer](ss::future<> f) {
```
This `finally` call has to become two `then_wrapped()` invocations due to wanting to return `writer->size_bytes()`... to maintain the status quo of exception handling, call `f.ignore_ready_future()` on the intermediate `future` from `reader.consume()` (Though I did add exception handling on the `writer->close()` here which wasn't handled previously, and goes against the inline comment :sweat:)

Maybe this would be nicer (and I believe equivalent) to write as: 

```cpp
    std::exception_ptr eptr;
    try {
        co_await reader.consume(
          index_filtered_copy_reducer(std::move(to_copy_index), *writer),
          model::no_timeout);
        writer->set_flag(compacted_index::footer_flags::self_compaction);
    } catch (...) {
        eptr = std::current_exception();
    }

    // do not handle exception on the close
    co_await writer->close();

    if (eptr) {
        std::rethrow_exception(eptr);
    }

    co_return writer->size_bytes();
```

WillemKauf committed:
`storage`: co-routinize `copy_filtered_entries()`

Exception handling should remain the same here in co-routinized version
due to `try/catch` blocks and capturing of exception with `eptr`.

There is one slight functional change here, which is that `writer->set_flag()`
is ONLY called if `reader.consume()` does not throw (which feels more correct
than setting the flag even in the presence of a thrown exception).

WillemKauf commented:
Force push to:
* Correct exception handling in new co-routinized versions of `write_clean_compacted_index()` and `copy_filtered_entries()` to match the previous behavior in continuation version.

WillemKauf commented:
I also wonder if this block here in `segment`:

https://github.com/redpanda-data/redpanda/blob/a4d5b01a4aacd2d13b24e737d4b9ba5bef28f252/src/v/storage/segment.cc#L123-L132

Would be better as

```cpp
    if (_appender) {
        u.data = _appender->file_byte_offset();
    } else if (_data_disk_usage_size.has_value()) {
        u.data = _data_disk_usage_size.value();
    } else {
        try {
            _data_disk_usage_size = co_await ss::file_size(
              _reader->path().string());
            u.data = _data_disk_usage_size.value();
        } catch (...) {
        }
    }
```

WillemKauf commented:
From another flamegraph https://pprof.me/c0b8a566d55645042947d909316939ed/ with compaction absent and a large number of segments:

I think even the changes here are insufficient in the absence of `compact` topics to reduce the number of syscalls.

I think all the places that we call into `clear_cached_disk_usage()` need review, and that we should make a best effort attempt to reset these values instead of clearing them.

I.e, `do_release_appender()`:

https://github.com/redpanda-data/redpanda/blob/9ca56e0141c65171dddc90cff6adc2eb5e5e9728/src/v/storage/segment.cc#L240-L251

This is a strong candidate for a rework, since we can now move the call site of `clear_cached_disk_usage()` knowing that `_idx.flush()` will set its cached disk use, and we can use the `file_byte_offset()` from both of our `segment_appender_ptr` and `compacted_index` to set the values via a call to `set_cached_disk_usage()`.

WillemKauf commented:
FYI reviewers, I checked out the flakey tests and none of the failures seem related to the changes.

WillemKauf commented on src/v/storage/spill_key_index.cc:
```diff
@@ -399,6 +399,13 @@ std::ostream& operator<<(std::ostream& o, const spill_key_index& k) {
 
+size_t spill_key_index::size_bytes() const {
+    if (_appender.has_value()) {
+        return _appender->file_byte_offset();
```
(FYI, this isn't a valid report of the file size for open `_appender`s, but since this PR only deals with closed `appender`s in the compaction layer, it isn't a worry... this is fixed in a follow up PR though, https://github.com/redpanda-data/redpanda/pull/26521)

dotnwat replied:
so would this be better if it returned like `optional<size>` and then the caller would do something like a stat since it didn't have a better source for the size?`

WillemKauf replied:
I think the lack of `appender` here actually indicates a `0` size.

oleiman commented on src/v/storage/segment_utils.cc:
```diff
@@ -268,13 +268,21 @@ ss::future<> write_clean_compacted_index(
+
+    if (eptr) {
+        std::rethrow_exception(eptr);
+    }
```
nitpick: is this a bit neater? your call

```suggestion
    // integrity verified in `do_detect_compaction_index_state`
    auto fut = co_await ss::coroutine::as_future(
        do_write_clean_compacted_index(reader, cfg, resources));
    co_await reader.close();
    if (fut.failed()) {
        std::rethrow_exception(fut.get_exception());
    }
```

WillemKauf replied:
Ah yeah, everything is cleaner with `as_future()` :)

oleiman commented on src/v/storage/types.h:
```diff
@@ -567,10 +567,14 @@ struct compaction_result {
+    compaction_result(
+      size_t before,
+      size_t after,
+      std::optional<size_t> cmp_idx_after = std::nullopt)
```
micro-nitpick: why not `cmp_idx_size_after`?

WillemKauf replied:
I forgot a word

oleiman commented on src/v/storage/segment_utils.cc:
```diff
@@ -258,20 +260,23 @@ static ss::future<> do_write_clean_compacted_index(
     try {
         // integrity verified in `do_detect_compaction_index_state`
-        co_await do_write_clean_compacted_index(reader, cfg, resources);
+        cmp_idx_size = co_await do_write_clean_compacted_index(
```
nit: pretty sure `compacted_index_reader::close` could be marked `noexcept`, so it seems like again `ss::coroutine::as_future` could make the scopes and whatnot a little easier to read.

WillemKauf replied:
Yeah, `as_future()` would be nicer here too, for a first swing at refactoring these functions in a non-functional way I figured `try/catch` would be easiest for reviewers to grok but I'm happy to change it

oleiman reviewed (commented):
looks pretty safe & sensible to me. 

> the changes here are insufficient in the absence of compact topics 

given that, and with all this plumbing, i can't help wondering whether there's some more general approach lurking in the shadows somewhere. don't have a concrete suggestion at the moment, but food for thought maybe

WillemKauf commented:
> given that, and with all this plumbing, i can't help wondering whether there's some more general approach lurking in the shadows somewhere

These changes are all squarely with compaction in mind- the plumbing is due to the call sites where we create, use and close `compacted_index_writer`s and `segment_appender`s. 

More general changes that serve to benefit the storage layer even in the absence of `compact` topics can be found here: https://github.com/redpanda-data/redpanda/pull/26521

This PR improves the situation by also relying on bytes reported by `writer`s/`appender`s as a replacement for a syscall from `ss::file_size`.

The plumbing/changes you see here are contained to the compaction subsystem, and I think with these two PRs combined the problem is more well handled from a general perspective than this PR in isolation.

dotnwat commented on src/v/storage/segment_index.cc:
```diff
@@ -229,6 +230,11 @@ ss::future<> segment_index::flush_to_file(ss::file backing_file) {
         co_await out.write(f.get(), f.size());
     }
+
+    if (!_disk_usage_size.has_value()) {
+        _disk_usage_size = b.size_bytes();
+    }
```
if we are truncating the file and writing out `b`, when would we ever want to use a cached value for the new file size? seems like it is _always_ going to be `b.size_bytes`

WillemKauf replied:
Fixed.

dotnwat commented on src/v/storage/segment_reader.cc:
```diff
@@ -51,8 +51,8 @@ segment_reader::~segment_reader() noexcept {
 
-    auto s = co_await stat();
-    set_file_size(s.st_size);
+    auto s = co_await fsize();
+    set_file_size(s);
```
👍

dotnwat replied:
did you find any interesting discussions online about stat vs seek for getting file size? i found some stuff that was pretty old form postgres mailing list, but hoping there was an updated source.

WillemKauf replied:
>did you find any interesting discussions online about stat vs seek for getting file size?

I tried to do some digging and found the same sources as you. Maybe @travisdowns has some opinions here. 

But also I'm wondering if `ss::file_size()` just calls into `fstat()` anyways... At the very least this change leads to a more direct statement.

dotnwat replied:
>But also I'm wondering if ss::file_size() just calls into fstat() anyways... At the very least this change leads to a more direct statement.

well, from what i see in seastar append_challenged_file uses a cached version of the file size already, and vanilla posix file impl uses lseek. with xfs (append challenged), as long as it isn't opened in read-only mode, then file_size() is probably already returning a cached value without using a system call.

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -268,13 +268,21 @@ ss::future<> write_clean_compacted_index(
-    // integrity verified in `do_detect_compaction_index_state`
-    return do_write_clean_compacted_index(reader, cfg, resources)
-      .finally([reader]() mutable {
-          return reader.close().then_wrapped(
-            [reader](ss::future<>) { /*ignore*/ });
```
nice this also would have risked creating ignored exceptional future error.

WillemKauf replied:
Yeah, this `/*ignore*/` isn't actually truthful!

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -268,13 +268,21 @@ ss::future<> write_clean_compacted_index(
+        eptr = std::current_exception();
+    }
+
+    co_await reader.close();
```
on the off chance that close failed, would we still want the eptr error which seems like it would probably be more interesting?

WillemKauf replied:
Hmm. not sure and don't have a terribly strong opinion, but I can alter this if you do.

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -228,23 +228,23 @@ ss::future<> copy_filtered_entries(
+        co_await reader.consume(
+          index_filtered_copy_reducer(std::move(to_copy_index), *writer),
+          model::no_timeout);
+        writer->set_flag(compacted_index::footer_flags::self_compaction);
```
>No functional changes.

in the pre-coroutine version, isn't the set_flag call happening in the finally block, so there is a chance it doesn't happen here if consume throws?

WillemKauf replied:
Oops right I guess this is a functional change.

Though, this now feels more correct. I will amend the commit message.

WillemKauf replied:
Leaving this as a `try/catch` block instead of using `ss::coroutine::as_future()` however, since both `reader.reset();` and `reader.consume()` could potentially throw, and wrapping both in a `try/catch` seems a bit more straightforward.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1661,7 +1661,8 @@ ss::future<> disk_log_impl::rewrite_segment_with_offset_map(
     co_await seg->index().drop_all_data();
```
> The major motivation for these changes is the number of calls being made
to `fstat()` for each `segment` during a housekeeping round. These syscalls
must go through the seastar `syscall_work_queue`, and take up much more
CPU utilization than they should when studying a flamegraph.


an interesting question is if switching to `::size()` instead of `::stat` would also fix the problem. afaict seastar doesn't sent `::size()` through the syscall workqueue, it just runs lseek directly inline on the reactor. not that that's good, but it might have the vsyscall optimization stuff.

still, having it cached is better in either case.

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -1203,9 +1212,15 @@ ss::future<compaction_result> concatenate_and_rebuild_target_segment(
+      cfg,
+      pb,
+      std::move(locks),
+      ret.cmp_idx_size_after.value());
```
in the context of this function, how do we know that `self_compact_segment` is returning `ret` with this size value set?

WillemKauf replied:
I figured it would be guaranteed because self compaction must occur for a new `segment` that has been created by an adjacent compaction (otherwise, the concatenated segment would be lacking the necessary in memory and on disk state) .

Just to be safe, I'll leave it optional at this level and through the call stack to `segment::set_cached_disk_usage()`.

dotnwat reviewed (commented):
looks good. +1 to using as_future helper. just a few questions/comments.

WillemKauf committed:
`storage`: use `ss::coroutine::as_future()` in `write_clean_compacted_index`

To make this block nicer.

WillemKauf commented:
Force push to:

* Use `std::optional<size_t> new_compacted_index_size` as argument for `segment::set_cached_disk_usage()`
* Remove `if()` condition in `segment_index::flush_to_file()` (we will now always set `_disk_usage_size = b.size_bytes();` here)
* Use `ss::coroutine::as_future()` in `write_clean_compacted_index()`
* rename `cmp_idx_after -> cmp_idx_size_after` in `compaction_result`

dotnwat commented on src/v/storage/segment.cc:
```diff
@@ -189,6 +189,14 @@ void segment::clear_cached_disk_usage() {
 
+void segment::set_cached_disk_usage(
+  size_t new_seg_size, std::optional<size_t> new_compacted_index_size) {
+    _data_disk_usage_size = new_seg_size;
+    if (new_compacted_index_size.has_value()) {
+        _compaction_index_size = new_compacted_index_size.value();
+    }
```
with the optional parameter, is the expectation that if the optional doesn't have a value that the last recorded size is the best estimate, or should it be reset to some default value?

WillemKauf replied:
> or should it be reset to some default value?

The assumption is that a call to `clear_cached_disk_usage()` precedes this call (which is true of the compaction context this PR is concerned with)

dotnwat commented on src/v/storage/segment_utils.cc:
```diff
@@ -258,23 +260,30 @@ static ss::future<> do_write_clean_compacted_index(
+    auto fut = co_await ss::coroutine::as_future(
+      do_write_clean_compacted_index(reader, cfg, resources));
+
+    co_await reader.close();
```
if `reader.close()` throws and `fut` failed, then `fut` will become an ignored exceptional future.

WillemKauf replied:
agh right, will fix

WillemKauf commented:
/backport v25.1.x

vbotbuildovich commented:
Failed to create a backport PR to v25.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-26513-v25.1.x-81 remotes/upstream/v25.1.x
git cherry-pick -x b6c34b1307 0fc0f8a56d 7e8f8cba2a 2323dad986 dfa021e8c9 321ed95b1a 23fb14a430 ce406e2035 1b69d44e62 754f03542e 9f8204cc3b 6a1f29363d 37727ee77e 8994ed77ca 207a790ba5
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/16010775616).
